pkg/controller/serviceaccount/serviceaccounts_controller.go

```diff
@@ -66,193 +67,159 @@ func DefaultServiceAccountsControllerOptions() ServiceAccountsControllerOptions
-func (e *ServiceAccountsController) namespaceUpdated(oldObj interface{}, newObj interface{}) {
+func (c *ServiceAccountsController) namespaceUpdated(oldObj interface{}, newObj interface{}) {
 	newNamespace := newObj.(*api.Namespace)
-	for _, sa := range e.serviceAccountsToEnsure {
-		e.createServiceAccountIfNeeded(sa, newNamespace.Name)
-	}
+	c.queue.Add(newNamespace.Name)
 }
 
-// createServiceAccountIfNeeded creates a ServiceAccount with the given name in the given namespace if:
-// * the named ServiceAccount does not already exist
-// * the specified namespace exists
-// * the specified namespace is in the ACTIVE phase
-func (e *ServiceAccountsController) createServiceAccountIfNeeded(sa api.ServiceAccount, namespace string) {
-	existingServiceAccount, err := e.getServiceAccount(sa.Name, namespace)
-	if err != nil {
-		glog.Error(err)
-		return
-	}
-	if existingServiceAccount != nil {
-		// If service account already exists, it doesn't need to be created
-		return
+func (c *ServiceAccountsController) runWorker() {
+	for c.processNextWorkItem() {
 	}
+}
 
-	namespaceObj, err := e.getNamespace(namespace)
-	if err != nil {
-		glog.Error(err)
-		return
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (c *ServiceAccountsController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
 	}
-	if namespaceObj == nil {
-		// If namespace does not exist, no service account is needed
-		return
-	}
-	if namespaceObj.Status.Phase != api.NamespaceActive {
-		// If namespace is not active, we shouldn't try to create anything
-		return
+	defer c.queue.Done(key)
+
+	err := c.syncHandler(key.(string))
+	if err == nil {
+		c.queue.Forget(key)
+		return true
 	}
 
-	e.createServiceAccount(sa, namespace)
-}
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
+	c.queue.AddRateLimited(key)
 
-// createDefaultServiceAccount creates a default ServiceAccount in the specified namespace
-func (e *ServiceAccountsController) createServiceAccount(sa api.ServiceAccount, namespace string) {
-	sa.Namespace = namespace
-	if _, err := e.client.Core().ServiceAccounts(namespace).Create(&sa); err != nil && !apierrs.IsAlreadyExists(err) {
-		glog.Error(err)
-	}
+	return true
 }
+func (c *ServiceAccountsController) syncNamespace(key string) error {
+	startTime := time.Now()
+	defer func() {
+		glog.V(4).Infof("Finished syncing namespace %q (%v)", key, time.Now().Sub(startTime))
+	}()
 
-// getServiceAccount returns the ServiceAccount with the given name for the given namespace
-func (e *ServiceAccountsController) getServiceAccount(name, namespace string) (*api.ServiceAccount, error) {
-	key := &api.ServiceAccount{ObjectMeta: api.ObjectMeta{Namespace: namespace}}
-	accounts, err := e.serviceAccounts.Index("namespace", key)
+	ns, err := c.nsLister.Get(key)
 	if err != nil {
-		return nil, err
+		return err
```

Comment:
do we need to tolerate NotFound here? should we add nsSynced to the WaitForCacheSync on startup?